synapse/storage/database.py

```diff
@@ -1154,21 +1153,11 @@ async def simple_upsert(
         requiring that a unique index exist on the column names used to detect a
         conflict (i.e. `keyvalues.keys()`).
 
-        If there is no such index, we can "emulate" an upsert with a SELECT followed
-        by either an INSERT or an UPDATE. This is unsafe: we cannot make the same
-        atomicity guarantees that a native upsert can and are very vulnerable to races
-        and crashes. Therefore if we wish to upsert without an appropriate unique index,
-        we must either:
-
-        1. Acquire a table-level lock before the emulated upsert (`lock=True`), or
-        2. VERY CAREFULLY ensure that we are the only thread and worker which will be
-           writing to this table, in which case we can proceed without a lock
-           (`lock=False`).
-
-        Generally speaking, you should use `lock=True`. If the table in question has a
-        unique index[*], this class will use a native upsert (which is atomic and so can
-        ignore the `lock` argument). Otherwise this class will use an emulated upsert,
-        in which case we want the safer option unless we been VERY CAREFUL.
+        If there is no such index yet[*], we can "emulate" an upsert with a SELECT
+        followed by either an INSERT or an UPDATE. This is unsafe: we cannot make the
+        same atomicity guarantees that a native upsert can and are very vulnerable to
+        races and crashes. Therefore if we wish to upsert without an appropriate unique
+        index, we must acquire a table-level lock before the emulated upsert.
```

Comment:
> we cannot make the same atomicity guarantees that a native upsert can

I wonder if my phrasing was poor here. Maybe we _can_ make those guarantees if we're using a sufficiently high isolation level?

Reply:
I tried to reword it a little. Suggestions welcome.